### DISCUSSION TITLE
[expo] Upgrading from Expo SDK 36 to 40 causing `undefined is not an object (evaluating '_window$location.protocol')`

### DISCUSSION THREAD

**Discussion by 5ervant** [vercel/next.js#21660] [Help] [answered]
**Summary**

Upgrading from SDK 36 to 40 with Next.js with Expo for Web causing `undefined is not an object (evaluating '_window$location.protocol')`.

**Environment**

  Expo CLI 4.1.3 environment info:
    System:
      OS: Windows 10 10.0.18363
    Binaries:
      Node: 15.7.0 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.16.0 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
      npm: 7.4.3 - C:\Program Files\nodejs\npm.CMD
      Watchman: 4.9.4 - C:\Users\USER\AppData\Local\watchman\watchman.EXE
    SDKs:
      Android SDK:
        API Levels: 26, 27, 28
        Build Tools: 26.0.3, 27.0.3, 28.0.3, 29.0.3
        System Images: android-26 | Google APIs Intel x86 Atom_64, android-26 | Google Play Intel x86 Atom, android-28 | Google APIs Intel x86 Atom
    IDEs:
      Android Studio: Version  3.6.0.0 AI-192.7142.36.36.6241897
    npmPackages:
      expo: ^40.0.0 => 40.0.1
      react: 16.13.1 => 16.13.1
      react-dom: 16.13.1 => 16.13.1
      react-native: https://github.com/expo/react-native/archive/sdk-40.0.1.tar.gz => 0.63.2
      react-native-web: ~0.13.12 => 0.13.18
    Expo Workflow: managed

**Please specify your device/emulator/simulator platform, model and version**

Galaxy A7 (2016) and Nexus_5X_API_26 (emulator)

**Error output**

Android (emulator):

> Exception
> undefined is not an object (evaluating '_window$location.protocol')

> C:\Users\USER\\...\PROJECT\node_modules\next\dist\next-server\lib\utils.js (3:193)

Console:

*Warning*
```
Unable to start your application. Please refer to https://expo.fyi/no-registered-application for more information.
- node_modules\react-native\Libraries\LogBox\LogBox.js:117:10 in registerWarning
- node_modules\react-native\Libraries\LogBox\LogBox.js:63:8 in warnImpl
- node_modules\react-native\Libraries\LogBox\LogBox.js:36:4 in console.warn
- node_modules\expo\build\environment\react-native-logs.fx.js:18:4 in warn
- node_modules\react-native\Libraries\ReactNative\AppRegistry.js:200:6 in runApplication
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:416:4 in __callFunction
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:109:6 in __guard$argument_0
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:364:10 in __guard
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:108:4 in callFunctionReturnFlushedQueue
* [native code]:null in callFunctionReturnFlushedQueue
```

*Error*
```
TypeError: undefined is not an object (evaluating '_window$location.protocol')
- node_modules\react-native\Libraries\LogBox\LogBox.js:148:8 in registerError
- node_modules\react-native\Libraries\LogBox\LogBox.js:59:8 in errorImpl
- node_modules\react-native\Libraries\LogBox\LogBox.js:33:4 in console.error
- node_modules\expo\build\environment\react-native-logs.fx.js:27:4 in error
- node_modules\react-native\Libraries\Core\ExceptionsManager.js:104:6 in reportException
- node_modules\react-native\Libraries\Core\ExceptionsManager.js:171:19 in handleException
- node_modules\react-native\Libraries\Core\setUpErrorHandling.js:24:6 in handleError
- node_modules\expo-error-recovery\build\ErrorRecovery.fx.js:9:32 in ErrorUtils.setGlobalHandler$argument_0
- node_modules\regenerator-runtime\runtime.js:63:36 in tryCatch
- node_modules\regenerator-runtime\runtime.js:293:29 in invoke
- node_modules\regenerator-runtime\runtime.js:63:36 in tryCatch
- node_modules\regenerator-runtime\runtime.js:154:27 in invoke
- node_modules\regenerator-runtime\runtime.js:164:18 in PromiseImpl.resolve.then$argument_0
- node_modules\react-native\node_modules\promise\setimmediate\core.js:37:13 in tryCallOne
- node_modules\react-native\node_modules\promise\setimmediate\core.js:123:24 in setImmediate$argument_0
- node_modules\react-native\Libraries\Core\Timers\JSTimers.js:130:14 in _callTimer
- node_modules\react-native\Libraries\Core\Timers\JSTimers.js:181:14 in _callImmediatesPass
- node_modules\react-native\Libraries\Core\Timers\JSTimers.js:441:30 in callImmediates
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:387:6 in __callImmediates
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:135:6 in __guard$argument_0
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:364:10 in __guard
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:134:4 in flushedQueue
* [native code]:null in flushedQueue
* [native code]:null in invokeCallbackAndReturnFlushedQueue
```

Web (browser):

> Failed to compile

> ./node_modules/react-native-elements/src/config/ViewPropTypes.js
> Module not found: Can't resolve 'react-native-web/dist/exports/ViewPropTypes' in 'C:\Users\USER\\...\PROJECT\node_modules\react-native-elements\src\config'

**Reproducible demo or steps to reproduce from a blank project**

- Have Expo SDK 36.
- Build a Next Expo app: https://docs.expo.io/guides/using-nextjs/
- Use Material-UI framework: https://material-ui.com/
- Upgrade from SDK 36 to 40.

---

Might help:

- [Raised `expo-cli` issue](https://github.com/expo/expo-cli/issues/3156)
- [Expo 39 and React navigation Error](https://forums.expo.io/t/expo-39-and-react-navigation-error/43152)
- [\**A lot of other packages are not able to be compiled after upgrading to react-native-web 0.12.x*](https://github.com/necolas/react-native-web/issues/1537#issuecomment-601167089)
I'm using 0.13.18.

**Comment by chrisdhanaraj** [comment, marked as answer]
@5ervant hey there! I ran into the same issue - I fixed this by upgrading my `next` package to the latest version. I went from Next 9 --> Next 10 and this resolved. For whatever reason, importing the `Router` object from `next/router` had the side-effect of checking `window.location.protocol`, even if the actual usage was wrapped.

**Reply by 5ervant** [reply]
How did you know that? Thanks, by the way.

**Reply by chrisdhanaraj** [reply]
Ah it was one of those process of elimination things. Essentially,

- I scoped my application down to the smallest component possible and got a small reproduction use-case
- Commented out the usage of `next/router`, saw that the error still happened
- Commented out the `import`, saw that the error stopped
- I've _personally_ run into some similar issues with my own packages with side-effects and double checked to see if a newer version of Next had fixed this issue - got lucky and it had!

If it hadn't, I'd have probably reached for a platform-specific file for any Router usage and gone from there, but thankful I didn't have to go too much deeper!
